apps/web/src/components/Basenames/UsernameProfileTransferOwnershipModal/context.tsx

```diff
@@ -0,0 +1,390 @@
+
+  const updateReclaim = useCallback(async () => {
+    if (!reclaimContract) return Promise.reject('Invalid reclaimContract');
+    await initiateReclaim(reclaimContract);
+  }, [initiateReclaim, reclaimContract]);
+
+  const updateSafeTransferFrom = useCallback(async () => {
+    if (!safeTransferFromContract) return Promise.reject('Invalid safeTransferFromContract');
+    await initiateSafeTransferFrom(safeTransferFromContract);
+  }, [initiateSafeTransferFrom, safeTransferFromContract]);
+
+  const updateSetName = useCallback(async () => {
+    if (!setNameContract) return Promise.reject('Invalid setNameContract');
+    await initiateSetName(setNameContract);
+  }, [initiateSetName, setNameContract]);
+
+  // Function & status we can track and display the edit rec
+  const ownershipSettings: OwnershipSettings[] = useMemo(() => {
+    return [
+      {
+        id: 'setAddr',
+        name: 'Address record',
+        description: 'Your Basename will resolve to this address.',
+        status: setAddrStatus,
+        contractFunction: updateSetAddr,
+      },
+      {
+        id: 'setName',
+        name: 'Name record',
+        description: 'Your Basename will no longer be displayed with your current address.',
+        status: setNameStatus,
+        contractFunction: updateSetName,
+      },
+      {
+        id: 'reclaim',
+        name: 'Profile editing',
+        description: 'Transfer editing rights to this address.',
+        status: reclaimStatus,
+        contractFunction: updateReclaim,
+      },
+      {
+        id: 'safeTransferFrom',
+        name: 'Token ownership',
+        description: 'Transfer the Basename token to this address.',
+        status: safeTransferFromStatus,
+        contractFunction: updateSafeTransferFrom,
+      },
+    ];
+  }, [
+    updateSetAddr,
+    updateSetName,
+    updateReclaim,
+    updateSafeTransferFrom,
+    reclaimStatus,
+    safeTransferFromStatus,
+    setAddrStatus,
+    setNameStatus,
+  ]);
+
+  const ownershipSettingsAreWaiting = useMemo(() => {
+    return ownershipSettings.every((ownershipSetting) => {
+      return [
+        WriteTransactionWithReceiptStatus.Idle,
+        WriteTransactionWithReceiptStatus.Success,
+      ].includes(ownershipSetting.status);
+    });
+  }, [ownershipSettings]);
+
+  useEffect(() => {
+    // Only when the wallet request steps is displaying
+    if (currentOwnershipStep !== OwnershipSteps.WalletRequests) return;
+
+    // Some transactions are loading / failed: return early
+    if (!ownershipSettingsAreWaiting) return;
+
+    // Smart wallet - we can batch calls
+    if (batchCallsEnabled) {
+      updateViaBatchCalls().catch((error) => {
+        setCurrentOwnershipStep(OwnershipSteps.OwnershipOverview);
+        logError(error, 'Failed to update via sendCalls');
+      });
+
+      return;
+    }
+
+    // For each ownership setting / wallet transaction
+    for (const ownershipSetting of ownershipSettings) {
```

Comment:
this feels like the meatiest part, eh?

Reply:
yeah, not sure it can be done better or with less lines. Recommend playing Hurt by Johnny Cash while reading through this part.